fe/fe-core/src/main/java/org/apache/doris/analysis/SelectStmt.java

```diff
@@ -603,14 +604,92 @@ public List<TupleId> collectTupleIds() {
         return result;
     }
 
-    private void whereClauseRewrite() {
+    private void rewriteWhereClause() {
         if (whereClause instanceof IntLiteral) {
             if (((IntLiteral) whereClause).getLongValue() == 0) {
                 whereClause = new BoolLiteral(false);
             } else {
                 whereClause = new BoolLiteral(true);
             }
         }
+
+        whereClause = inferIsNotNull(whereClause);
+    }
+
+    /*-
+     * Infer `IS NOT NULL` from predicate in whereClause.
+     * And spilt it into two parts.
+     * Example:
+     * Where col > 0; -> Where col > 0 and col is not null;
+     */
+    private Expr inferIsNotNull(Expr expr) {
```

Comment:
It should be a new Rule ~

Reply:
Note: It isn't a `optimizer rule` 

It is because the following transformation can't terminate in Finite State Machine of rule.

```sql
select * from t1 where c1 > 0;  
-> 
select * from t1 where c1 > 0 and c1 is not null;
->
select * from t1 where c1 > 0 and c1 is not null and c1 is not null;
->
.....
```

So, In fact, this is a `rewrite` in planner build.

Reply:
What' the difference between rule and rewrite?

Reply:
This key point is `in planner build.`
Current rule is recursively optimized.

Reply:
We should make sure that a `SlotRef` has only one is-not-null predicate, no matter the infer-not-null logical is in a recursively applied rule or single-pass plan rewrite.

Please check the following cases:
1. There is already an is-not-null predicate, e.g., `SELECT a FROM t WHERE a > 1 AND a is not NULL`.
2. More than one where clauses refer to the same `SlotRef` that could infer an is-not-null predicate, e.g., `SELECT * FROM (SELECT * FROM t WHERE a > 1) t WHERE a < 2`.

One possible solution is that we track the relationship between a `SlotRef` and its is-not-null predicate in `Analyzer`.